gnc/sub8_perception/src/sub8_pcl/cv_utils.cc

```diff
@@ -1,12 +1,270 @@
+		if(abs_slope < low_abs_derivative_thresh){ 
+			low_thresh = i;
+			break; 
+		}
+	}
+	// std::cout << "low_thresh = " << low_thresh << std::endl;
+	// std::cout << std::endl;
+	std::string ros_log = 
+		( boost::format("Target: %1%\nClosest distribution mode: %2%  Thresholds selected:  low=%3%  high=%4%")
+		% target % target_mode % low_thresh % high_thresh ).str() ;
+	ROS_INFO(ros_log.c_str());
+
+	// Threshold image
+	cv::inRange(src, low_thresh, high_thresh, dest);
+
+	#ifdef VISUALIZE
+	// Prepare to draw graph of histogram and derivative
+	int hist_w = 512; int hist_h = 400;
+	int bin_w = cvRound( (double) hist_w/hist_size );
+	cv::Mat histImage( hist_h, hist_w, CV_8UC1, cv::Scalar( 0,0,0) );
+	cv::Mat histDerivImage( hist_h, hist_w, CV_8UC1, cv::Scalar( 0,0,0) );
+	cv::normalize(hist_smooth, hist_smooth, 0, histImage.rows, cv::NORM_MINMAX, -1, cv::Mat() );
+	cv::normalize(hist_derivative, hist_derivative, 0, histImage.rows, cv::NORM_MINMAX, -1, cv::Mat() );
+
+	// Draw Graphs
+	for( int i = 1; i < hist_size; i++ )
+	{
+	// Plot image histogram
+	cv::line( histImage, cv::Point( bin_w*(i-1), hist_h - cvRound(hist_smooth.at<float>(i-1)) ) ,
+	                 cv::Point( bin_w*(i), hist_h - cvRound(hist_smooth.at<float>(i)) ),
+	                 cv::Scalar( 255, 0, 0), 2, 8, 0  );
+	// Plot image histogram derivative
+	cv::line( histDerivImage, cv::Point( bin_w*(i-1), hist_h - cvRound(hist_derivative.at<float>(i-1)) ) ,
+	                 cv::Point( bin_w*(i), hist_h - cvRound(hist_derivative.at<float>(i)) ),
+	                 cv::Scalar( 122, 0, 0), 1, 8, 0  );
+	}
+
+	// Shade in area being segmented under histogram curve
+	cv::line( histImage, cv::Point( bin_w*low_thresh, hist_h - cvRound(hist_smooth.at<float>(low_thresh)) ) ,
+	                 cv::Point( bin_w*low_thresh, hist_h), cv::Scalar( 125, 0, 0),2);
+	cv::line( histImage, cv::Point( bin_w*high_thresh, hist_h - cvRound(hist_smooth.at<float>(high_thresh)) ) ,
+                 cv::Point( bin_w*high_thresh, hist_h), cv::Scalar( 125, 0, 0),2);
+	cv::floodFill(histImage, cv::Point(bin_w*cvRound(float(low_thresh + high_thresh) / 2.0), hist_h - 1), cv::Scalar(125));
+
+	// Combine graphs into one image and display results
+	cv::Mat segmentation_channel = cv::Mat::zeros(histImage.size(), CV_8UC1);
+	cv::Rect upper_right_corner = cv::Rect(histImage.cols - dest.cols - 1, 0, dest.cols, dest.rows);
+	dest.copyTo(segmentation_channel(upper_right_corner));
+	segmentation_channel = segmentation_channel * 0.3;
+	std::vector<cv::Mat> debug_img_channels;
+	debug_img_channels.push_back(histImage); 
+	debug_img_channels.push_back(histDerivImage);
+	debug_img_channels.push_back(segmentation_channel);
+	cv::Mat debug_img;
+	cv::merge(debug_img_channels, debug_img);
+	cv::imshow((boost::format("Statistical Image Segmentation(%1%)") % image_name).str(), debug_img );
```

Comment:
Oh - that is quite a hidden ifdef. Preprocessor directives should be the start of the line, so the if/endif is clear